conda/match.py

```diff
@@ -0,0 +1,203 @@
+
+            elif isinstance(spec, string_types):
+                spec, _, oparts = spec.partition('(')
+                parts = spec.strip().split()
+                assert 1 <= len(parts) <= 3, repr(spec)
+                name, version, build = (parts + ['*', '*'])[:3]
+                push_(('name', name), ('version', version), ('build', build))
+                if normalize and (self.exact_field('version') is not None and
+                                  self.exact_field('build') is None and
+                                  self.exact_field('name') is not None):
+                    # When someone supplies 'foo=a.b' on the command line, we
+                    # want to append an asterisk; e.g., 'foo=a.b.*', but only
+                    # if there is not also an exact build and name. In that
+                    # case we assume the user is looking for an exact match.
+                    ver = self.exact_field('version')
+                    self.specs['version'] = VersionSpec(ver + '*')
+                if oparts:
+                    if oparts.strip()[-1] != ')':
+                        raise CondaValueError("Invalid MatchSpec: %s" % spec)
+                    for opart in oparts.strip()[:-1].split(','):
+                        field, eq, value = (x.strip() for x in opart.partition('='))
+                        if not field:
+                            continue
+                        elif not value and (eq or field != 'optional'):
+                            raise CondaValueError("Invalid MatchSpec: %s" % spec)
+                        elif field == 'optional':
+                            kwargs.setdefault('optional', bool(value) if eq else True)
+                        elif field == 'target':
+                            kwargs.setdefault('target', value)
+                        else:
+                            push_((field, literal_eval(value)))
+
+            else:
+                raise CondaValueError("Cannot construct MatchSpec from: %r" % (spec,))
+
+        self.target = kwargs.pop('target', None)
+        self.optional = bool(kwargs.pop('optional', False))
+        push_(*iteritems(kwargs))
+        return self
+
+    def exact_field(self, fn):
+        v = self.specs.get(fn)
+        return None if v is None or hasattr(v, 'match') else v
+
+    def is_exact(self):
+        return all(self.exact_field(x) is not None for x in ('fn', 'schannel'))
+
+    def is_simple(self):
+        return len(self.specs) == 1 and self.exact_field('name') is not None
+
+    def match(self, rec):
+        # type: (IndexRecord) -> bool
+        for f, v in iteritems(self.specs):
+            val = getattr(rec, f)
+            if not (v.match(val) if hasattr(v, 'match') else v == val):
+                return False
+        return True
+
+    def to_filename(self):
+        tmp = self.exact_field('fn')
+        if tmp:
+            return tmp
+        vals = [self.exact_field(x) for x in ('name', 'version', 'build')]
```

Comment:
make this a tuple comprehension, and then line 129 can be just

    return '%s-%s-%s.tar.bz2' % vals